### ISSUE TITLE
AutoML Wrapper in Deepchem

### ISSUE THREAD

**Issue by Milan933-coder** [deepchem/deepchem#4909]
# 🚀 Feature Request: AutoML Wrapper for DeepChem with Optuna Integration

## 📌 Summary
DeepChem currently provides a rich ecosystem for molecular machine learning, including datasets, featurizers, and model implementations. However, it lacks a unified AutoML framework for:
- Automated model selection
- Hyperparameter optimization
- Cross-validation workflows

This proposal introduces an AutoML wrapper that integrates Optuna to streamline regression and classification tasks.

---

## 🎯 Motivation
In real-world molecular ML workflows, users often:
- Manually try multiple models (GraphConv, MPNN, AttentiveFP, etc.)
- Tune hyperparameters via trial-and-error
- Write repetitive boilerplate for evaluation

This leads to:
- ❌ Inefficient experimentation
- ❌ Poor reproducibility
- ❌ High barrier for new users

An AutoML layer would:
- ✅ Standardize experimentation
- ✅ Improve performance via systematic tuning
- ✅ Reduce development time significantly

---

## 💡 Proposed Solution
Introduce a module: `deepchem.automl`

### Core Features
1. **Model Zoo Integration**
   - Support multiple DeepChem models:
     - GraphConvModel
     - MPNNModel
     - AttentiveFPModel
     - ChemBERTa (via wrappers)
     - Custom user-defined models

2. **Optuna-based Hyperparameter Optimization**
   - Seamless integration with Optuna
   - Define search spaces per model
   - Support:
     - Learning rate
     - Batch size
     - Hidden dimensions
     - Number of layers
     - Dropout rates

3. **Task Support**
   - Regression (e.g., solubility, FFV, Tg)
   - Classification (e.g., toxicity, activity)

4. **Unified API**
   Example usage:
   ```python
   from deepchem.automl import AutoML

   automl = AutoML(
       task_type="regression",
       models=["graphconv", "mpnn", "attentivefp"],
       n_trials=50,
       metric="r2",
   )

   automl.fit(train_dataset, valid_dataset)

   best_model = automl.best_model
   results = automl.evaluate(test_dataset)
   ```

5. **Cross-Validation Support**
   - K-Fold CV integration
   - Stratified splitting (for classification)
   - Scaffold splitting (important for chemistry)

6. **Experiment Tracking**
   - Track:
     - Trials
     - Metrics
     - Hyperparameters
   - Optional integration:
     - TensorBoard
     - Weights & Biases

---

## 🧱 Suggested Architecture
```plaintext
deepchem/
 └── automl/
      ├── automl.py          # Main interface
      ├── search_space.py    # Hyperparameter definitions
      ├── trainer.py         # Training logic
      ├── evaluator.py       # Metrics + validation
      └── utils.py
```

---

## 🔬 Example: Optuna Objective
```python
def objective(trial):
    model_type = trial.suggest_categorical("model", ["graphconv", "mpnn"])

    if model_type == "graphconv":
        lr = trial.suggest_loguniform("lr", 1e-5, 1e-2)
        model = GraphConvModel(n_tasks=1, learning_rate=lr)

    model.fit(train_dataset)

    score = model.evaluate(valid_dataset, [metric])["r2"]
    return score
```

---

## 🌍 Impact
This feature would:
- Make DeepChem more competitive with libraries like AutoGluon / AutoSklearn
- Attract new users (especially beginners in ML for chemistry)
- Enable faster research iteration cycles
- Standardize benchmarking workflows

---

## 🔥 Bonus Extensions (Future Work)
- Neural Architecture Search (NAS) for GNNs
- Meta-learning for dataset-specific tuning
- Distributed hyperparameter optimization
- Integration with Ray Tune

---

## 🙌 Contribution
I would be happy to:
- Contribute to the initial implementation
- Design modular APIs
- Add examples and benchmarks
- Already generated a PR for OLMo integration in hf_models
- (https://github.com/deepchem/deepchem/pull/4848)
---

## 📎 Closing Note
This feature aligns strongly with DeepChem’s vision of making AI for drug discovery more accessible and reproducible.

An AutoML wrapper would significantly enhance usability and research productivity.

**Comment by Milan933-coder** [comment]
@ARY2260  @riya-singh28  Can i work on this
